containerize/command.go

```diff
@@ -0,0 +1,290 @@
+package containerize
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"os/user"
+	"runtime"
+	"strconv"
+	"sync"
+	"syscall"
+	"teleport-exec/cgroups"
+	"teleport-exec/filestream"
+
+	"github.com/google/uuid"
+	"go.uber.org/multierr"
+)
+
+//-------------------------------------------------------------------------------------------------
+type Command struct {
+	CommandId string
+	Command   string
+	Args      []string
+	executor  *exec.Cmd
+	log       *ProcessLog
+
+	mu sync.RWMutex
+
+	cgroup *cgroups.Container
+
+	done chan bool // Used to notify all Wait() calls that the process has died
+
+	running  bool // Used to easily check the running status of the process without waitpid, etc
+	isClosed bool // Used to prevent double-close
+	started  bool // Used to prevent double-start
+}
+
+//-------------------------------------------------------------------------------------------------
+func NewCommand(command []string) (*Command, error) {
+	c := Command{
+		CommandId: uuid.NewString(),
+		Command:   command[0],
+		Args:      command[1:],
+		running:   false,
+		done:      make(chan bool),
+	}
+
+	// Set up the command execution
+	cmd := exec.Command(c.Command, c.Args...)
+	c.executor = cmd
+	cmd.SysProcAttr = c.sysProcAttr()
+	cmd.Env = []string{
+		"TERM=vt100",
+	}
+
+	// Redirect command output to a command-specific log (so that we could read/stream it later)
+	pl, err := NewProcessLog(c.CommandId)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set up logging for command '%s': %w", c.CommandId, err)
+	}
+	c.log = pl
+	cmd.Stdout = pl.fd
+	cmd.Stderr = pl.fd
+
+	return &c, nil
+}
+
+//-------------------------------------------------------------------------------------------------
+// Starts the command in a separate thread
+func (c *Command) Start() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.started {
+		return errors.New("command is already running")
+	}
+	c.started = true
+
+	// On Linux, pdeathsig will kill the child process when the thread dies,
+	// not when the process dies. runtime.LockOSThread ensures that as long
+	// as this function is executing that OS thread will still be around
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
```

Comment:
but `waitForCompletion` will run in a separate goroutine, is there any point to this?

Reply:
Updated the comment to clarify what's happening here